src/Features/Core/Portable/ConvertAnonymousTypeToClass/AbstractConvertAnonymousTypeToClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,414 @@
+            // properly replace them even in that case.
+            editor.ReplaceNode(
+                childCreation,
+                (currentNode, g) =>
+                {
+                    var currentAnonymousObject = (TAnonymousObjectCreationExpressionSyntax)currentNode;
+
+                    // If we hit the node the user started on, then add the rename annotation here.
+                    var classNameToken = startingCreationNode == childCreation
+                        ? g.Identifier(className).WithAdditionalAnnotations(RenameAnnotation.Create())
+                        : g.Identifier(className);
+
+                    var classNameNode = (TIdentifierNameSyntax)g.IdentifierName(classNameToken);
+                    return CreateObjectCreationExpression(classNameNode, currentAnonymousObject)
+                        .WithAdditionalAnnotations(Formatter.Annotation);
+                });
+        }
+
+        private static async Task<INamedTypeSymbol> GenerateFinalNamedTypeAsync(
+            Document document, string className,
+            ImmutableArray<IPropertySymbol> properties, CancellationToken cancellationToken)
+        {
+            var compilation = await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+            // Fext, see if any of the properties ended up using any type parameters from the
+            // containing method/named-type.  If so, we'll need to generate a generic type so we can
+            // properly pass these along.
+            var capturedTypeParameters =
+                properties.Select(p => p.Type)
+                          .SelectMany(t => t.GetReferencedTypeParameters())
+                          .Distinct()
+                          .ToImmutableArray();
+
+            // Now try to generate all the members that will go in the new class. This is a bit
+            // circular.  In order to generate some of the members, we need to know about the type.
+            // But in order to create the type, we need the members.  To address this we do two
+            // passes. First, we create an empty version of the class.  This can then be used to
+            // help create members like Equals/GetHashCode.  Then, once we have all the members we
+            // create the final type.
+            var namedTypeWithoutMembers = CreateNamedType(className, capturedTypeParameters, members: default);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var constructor = CreateConstructor(compilation, className, properties, generator);
+
+            // Generate Equals/GetHashCode.  Only readonly properties are suitable for these
+            // methods.  We can defer to our existing language service for this so that we
+            // generate the same Equals/GetHashCode that our other IDE features generate.
+            var readonlyProperties = ImmutableArray<ISymbol>.CastUp(
```

Comment:
>readonlyProperties [](start = 16, length = 18)

What's a scenario where some properties get filtered out? Do we have a test for it?

Reply:
For mutable (settable) properties.  This happens in VB with non-Key anonymous-type properties.  I have tests for this.